pkg/engine/constraints.go

```diff
@@ -0,0 +1,127 @@
+package engine
+
+import (
+	"fmt"
+	"os"
+
+	"github.com/klothoplatform/klotho/pkg/core"
+	"github.com/klothoplatform/klotho/pkg/graph"
+	"gopkg.in/yaml.v3"
+)
+
+type (
+	Constraint interface {
+		Level() ConstraintLevel
+		IsSatisfied(dag *core.ResourceGraph) bool
+		CanApply() bool
+		Apply(dag *core.ResourceGraph) error
+	}
+	// [
+	// 	{level: "application", operator: "must_exist", TargetNode.... TargetEdge.... Resource....}
+	// 	{
+	// 		level: "application",
+	// 		operator: "must_exist",
+	// 		data: {
+	// 			TargetEdge: edfe.....core
+	// 			Node: Resource....
+	// 		}
+	// 	}
+	// ]
+	//must contain
+	//must not exist ...
+
+	BaseConstraint struct {
+		Level    ConstraintLevel    `yaml:"level"`
+		Operator ConstraintOperator `yaml:"operator"`
+	}
+
+	// Add Remove Rename      Replace(Maybe defer)
+	ApplicationConstraint struct {
+		Operator ConstraintOperator            `yaml:"operator"`
+		Node     core.ResourceId               `yaml:"node"`
+		Edge     []graph.Edge[core.ResourceId] `yaml:"edge"`
+	}
+
+	// Add execunit1
+	// application must_contain node of lambda
+
+	ConstructConstraint struct {
+		Operator   ConstraintOperator `yaml:"operator"`
+		Target     core.ResourceId    `yaml:"target"`
+		Type       string             `yaml:"type"`
+		Attributes map[string]any     `yaml:"attributes"`
+	}
+
+	EdgeConstraint struct {
+		Operator ConstraintOperator            `yaml:"operator"`
+		Target   []graph.Edge[core.ResourceId] `yaml:"target"`
+		Node     core.ResourceId               `yaml:"node"`
+	}
+
+	// NodeConstraint struct {
+	// 	BaseConstraint
+	// 	Target   core.ResourceId
+	// 	Property string
+	// 	Value    any
+	// }
+
+	ConstraintLevel    string
```

Comment:
What kind of values will this be? Is it enum-like (e.g. `high`, `low`, `required`) or arbitrary?

Edit: oh, I see in the example below. I'd suggest calling this a scope, not level. Level is ambiguous, and I initially thought this was a _priority_ level.

Reply:
scope makes sense to me, definitely better naming